modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockHttpProxyServer.java

```diff
@@ -7,91 +7,69 @@
  */
 package org.elasticsearch.repositories.gcs;
 
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
-import org.elasticsearch.common.network.NetworkAddress;
-import org.elasticsearch.mocksocket.MockServerSocket;
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpObjectAggregator;
+import io.netty.handler.codec.http.HttpServerCodec;
 
-import java.io.BufferedInputStream;
-import java.io.BufferedOutputStream;
 import java.io.Closeable;
 import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.net.Socket;
-import java.net.SocketException;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
+import java.net.InetSocketAddress;
+import java.util.function.Supplier;
 
 /**
  * A mock HTTP Proxy server for testing of support of HTTP proxies in various SDKs
  */
 class MockHttpProxyServer implements Closeable {
 
-    private static final Logger log = LogManager.getLogger(MockHttpProxyServer.class);
+    private final EventLoopGroup bossGroup = new NioEventLoopGroup(1);
+    private final EventLoopGroup workerGroup = new NioEventLoopGroup();
 
-    private final MockServerSocket serverSocket;
-    private final Thread serverThread;
-    private final CountDownLatch latch;
-    private final ExecutorService executorService = Executors.newCachedThreadPool();
+    private ChannelFuture channelFuture;
+    private InetSocketAddress socketAddress;
 
-    MockHttpProxyServer(SocketRequestHandler handler) throws IOException {
-        // Emulate a proxy HTTP server with plain sockets because MockHttpServer doesn't work as a proxy
```

Comment:
Could you elaborate why we can't use our normal approach to mock http servers here? It doesn't seem like we're doing anything special in these handlers that would prevent us from using it?

Reply:
The problem is that a normal `com.sun.net.httpserver.HttpServer` can't act as a proxy server, it just doesn't accept requests. E.g. this test fails:
```java
HttpServer proxyServer = MockHttpServer.createHttp(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 0);
proxyServer.createContext("/", exchange -> {
    byte[] bytes = httpBody.getBytes(StandardCharsets.UTF_8);
    exchange.getResponseBody().write(bytes);
    exchange.sendResponseHeaders(200, 0);
    exchange.close();
});
var httpClient = HttpClients.custom()
    .setRoutePlanner(
        new DefaultProxyRoutePlanner(new HttpHost(InetAddress.getLoopbackAddress(), proxyServer.getAddress().getPort()))
    )
    .build();
try (
    httpClient;
    var httpResponse = SocketAccess.doPrivilegedIOException(() -> httpClient.execute(new HttpGet("http://googleapis.com/")))
) {
    assertEquals(httpBody.length(), httpResponse.getEntity().getContentLength());
    assertEquals(httpBody, EntityUtils.toString(httpResponse.getEntity()));
}